gadget/install/install_test.go

```diff
@@ -1165,3 +1166,215 @@ func (s *installSuite) TestInstallEncryptPartitionsNoDeviceSet(c *C) {
+		return nil
+	})
+	defer restore()
+
+	gadgetRoot := filepath.Join(c.MkDir(), "gadget")
+	ginfo, _, _, restore, err := gadgettest.MockGadgetPartitionedDisk(gadgettest.SingleVolumeUC20GadgetYaml, gadgetRoot)
+	c.Assert(err, IsNil)
+	defer restore()
+
+	// Fill in additional information about the target device as the installer does
+	partIdx := 1
+	for i, part := range ginfo.Volumes["pc"].Structure {
+		if part.Role == "mbr" {
+			continue
+		}
+		ginfo.Volumes["pc"].Structure[i].Device = "/dev/vda" + strconv.Itoa(partIdx)
+		partIdx++
+	}
+	// Fill encrypted partitions if encrypting
+	var esd *install.EncryptionSetupData
+	if opts.encryption {
+		labelToEncData := map[string]*install.MockEncryptedDeviceAndRole{
+			"ubuntu-save": {
+				Role:            "system-save",
+				EncryptedDevice: "/dev/mapper/ubuntu-save",
+			},
+			"ubuntu-data": {
+				Role:            "system-data",
+				EncryptedDevice: "/dev/mapper/ubuntu-data",
+			},
+		}
+		esd = install.MockEncryptionSetupData(labelToEncData)
+	}
+
+	// 10 million mocks later ...
+	// finally actually run MountVolumes
+	seedMntDir, unmount, err := install.MountVolumes(ginfo.Volumes, esd)
+	c.Assert(err, IsNil)
+	c.Assert(seedMntDir, Equals, seedMntPt)
+
+	err = unmount()
+	c.Assert(err, IsNil)
+
+	c.Assert(mountCall, Equals, 4)
+	c.Assert(umountCall, Equals, 4)
+}
+
+func (s *installSuite) TestMountVolumesSimpleHappy(c *C) {
+	s.testMountVolumes(c, mountVolumesOpts{
+		encryption: false,
+	})
+}
+
+func (s *installSuite) TestMountVolumesSimpleHappyEncrypted(c *C) {
+	s.testMountVolumes(c, mountVolumesOpts{
+		encryption: true,
+	})
+}
+
+func (s *installSuite) TestMountVolumesZeroSeeds(c *C) {
+	onVolumes := map[string]*gadget.Volume{}
+	_, _, err := install.MountVolumes(onVolumes, nil)
+	c.Assert(err, ErrorMatches, "there are 0 system-seed{,-null} partitions, expected one")
+}
+
+func (s *installSuite) TestMountVolumesManySeeds(c *C) {
+	onVolumes := map[string]*gadget.Volume{
+		"pc": {
+			Structure: []gadget.VolumeStructure{
+				{Name: "system-seed", Filesystem: "vfat", Role: gadget.SystemSeed},
+				{Name: "system-seed-null", Filesystem: "vfat", Role: gadget.SystemSeedNull},
+			},
+		},
+	}
+
+	mountCall := 0
+	restore := install.MockSysMount(func(source, target, fstype string, flags uintptr, data string) error {
+		mountCall++
+		c.Assert(flags, Equals, uintptr(0))
+		return nil
+	})
+	defer restore()
+
+	umountCall := 0
+	restore = install.MockSysUnmount(func(target string, flags int) error {
+		umountCall++
+		c.Assert(flags, Equals, 0)
+		return nil
+	})
+	defer restore()
+
+	_, _, err := install.MountVolumes(onVolumes, nil)
+	c.Assert(err, ErrorMatches, "there are 2 system-seed{,-null} partitions, expected one")
+
+	c.Assert(mountCall, Equals, 2)
+	// check unmount is called implicitly on error for cleanup
+	c.Assert(umountCall, Equals, 2)
+}
+
+func (s *installSuite) TestMountVolumesLazyUnmount(c *C) {
+	seedMntPt := filepath.Join(s.dir, "run/mnt/ubuntu-seed")
```

Comment:
Bonus points if you could do additionally:
```go
log, restore := logger.MockLogger()
defer restore()
...
c.Check(log.String(), Contains, "cannot umount ....: forcing lazy umount (trying lazy umount next")
```

Reply:
Updated, Thanks!

Reply:
a side note, I noticed that the [fallback logic](https://github.com/snapcore/snapd/blob/85cdda5c732c6e6de04f522a5b076badb0c7b393/gadget/install/install.go#L576) will always log a problem even if fallback works. It's out of the scope of this PR, but I thought I mention it. It should already be fixed in this [PR](https://github.com/snapcore/snapd/pull/12967/files).